File: lab4.py
import pandas as pd
import numpy as np
from flask import Flask, request, jsonify

from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

# Load the dataset
url = 'https://raw.githubusercontent.com/Seun999/datasets/main/Fish.csv'
df = pd.read_csv(url)

# Check for missing values
df.dropna(inplace=True)

# One-hot encode the 'Species' column
df = pd.get_dummies(df, columns=['Species'])

# Split the data into features (X) and the target variable (y)
X = df.drop('Weight', axis=1)
y = df['Weight']

# Create and train the regression model
model = LinearRegression()
model.fit(X, y)

# Create the Flask app
app = Flask(__name__)

# Define the prediction route
@app.route('/predict', methods=['POST'])
def predict():
    data = request.get_json()

    # Extract the input features from the JSON data
    species = data['Species']
    length1 = data['Length1']
    length2 = data['Length2']
    length3 = data['Length3']
    height = data['Height']
    width = data['Width']

    # Encode the species input using one-hot encoding
    species_bream = 1 if species == 'Bream' else 0
    species_parkki = 1 if species == 'Parkki' else 0
    species_perch = 1 if species == 'Perch' else 0
    species_pike = 1 if species == 'Pike' else 0
    species_roach = 1 if species == 'Roach' else 0
    species_smelt = 1 if species == 'Smelt' else 0
    species_whitefish = 1 if species == 'Whitefish' else 0

    # Prepare the input data for prediction
    input_data = np.array([[length1, length2, length3, height, width,
                            species_bream, species_parkki, species_perch,
                            species_pike, species_roach, species_smelt,
                            species_whitefish]])

    # Perform the prediction
    prediction = model.predict(input_data)

    # Create a response JSON with the prediction result
    response = {
        'predicted_weight': prediction[0]
    }

    return jsonify(response)

# Run the app
if __name__ == '__main__':
    app.run(debug=True)

File: Lab 4.js
document.getElementById("fish-form").addEventListener("submit", function(event) {
    event.preventDefault();

    const formData = new FormData(event.target);
    const species = formData.get("species");
    const length1 = parseFloat(formData.get("length1"));
    const length2 = parseFloat(formData.get("length2"));
    const length3 = parseFloat(formData.get("length3"));
    const height = parseFloat(formData.get("height"));
    const width = parseFloat(formData.get("width"));

    // Encode the species input using one-hot encoding
    const speciesBream = species === "Bream" ? 1 : 0;
    const speciesParkki = species === "Parkki" ? 1 : 0;
    const speciesPerch = species === "Perch" ? 1 : 0;
    const speciesPike = species === "Pike" ? 1 : 0;
    const speciesRoach = species === "Roach" ? 1 : 0;
    const speciesSmelt = species === "Smelt" ? 1 : 0;
    const speciesWhitefish = species === "Whitefish" ? 1 : 0;

    const inputData = {
        "Length1": length1,
        "Length2": length2,
        "Length3": length3,
        "Height": height,
        "Width": width,
        "Species_Bream": speciesBream,
        "Species_Parkki": speciesParkki,
        "Species_Perch": speciesPerch,
        "Species_Pike": speciesPike,
        "Species_Roach": speciesRoach,
        "Species_Smelt": speciesSmelt,
        "Species_Whitefish": speciesWhitefish
    };

    // Send the input data to the Flask API for prediction
    fetch('http://localhost:5000/predict', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(inputData)
    })
    .then(response => response.json())
    .then(data => {
        // Display the prediction result in the "prediction-result" div
        document.getElementById('prediction-result').textContent = "Predicted Weight: " + data.predicted_weight;
    })
    .catch(error => console.error('Error:', error));
});
